internal/wait/wait.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the Apache License, Version 2.0, see LICENCE file for details.
+
+package wait
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/retry"
+	"github.com/juju/terraform-provider-juju/internal/juju"
+)
+
+// Logger is an interface for logging debug messages.
+type Logger interface {
+	Debugf(msg string, additionalFields ...map[string]interface{})
+}
+
+// RetryConf is a struct to configure the retry behavior.
+type RetryConf struct {
+	// MaxDuration is the maximum duration to wait for the condition to be met.
+	MaxDuration time.Duration
+	// Delay is the delay between retries.
+	Delay time.Duration
+	// Clock is the clock to use for timing.
+	Clock clock.Clock
+}
+
+// WaitForCfg is a configuration structure for the WaitFor function.
+type WaitForCfg[I any, D any] struct {
+	Context context.Context
+
+	// GetData is a function that retrieves data based on the input.
+	GetData GetData[I, D]
+	// Input is the input to be passed to the GetData function.
+	Input I
+	// DataAssertions is a list of assertions to check the data against.
+	// If any assertion fails, the function will return an error.
+	DataAssertions []Assert[D]
+	// NonFatalErrors is a list of non-fatal errors to ignore.
+	NonFatalErrors []error
```

Comment:
what if instead of `NonFatalErrors` we passed in a function that would tell us if the error is a fatal error? e.g.
`IsFatalError func(err error) bool`